### PR TITLE
Cleanup static table definitions to be more concise and typesafe.

### PR DESCRIPTION
**Pull request by mkleen**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/metadata/table/ColumnRegistrar.java [hunk 1]
```diff
@@ -54,17 +57,44 @@ public ColumnRegistrar(RelationName relationName, RowGranularity rowGranularity)
         this.rowGranularity = rowGranularity;
         this.infosBuilder = ImmutableSortedMap.naturalOrder();
         this.columnsBuilder = ImmutableSortedSet.orderedBy(Reference.COMPARE_BY_COLUMN_IDENT);
+        this.expressionBuilder = ImmutableSortedMap.naturalOrder();
     }
 
-    public ColumnRegistrar register(String column, DataType type) {
-        return register(new ColumnIdent(column), type);
+    public ColumnRegistrar<T> register(String column, DataType type) {
```
Comment (kovrus) [inline sql/src/main/java/io/crate/metadata/table/ColumnRegistrar.java]: should DataType be generic here and in two methods bellow?

Comment (mkleen) [reply sql/src/main/java/io/crate/metadata/table/ColumnRegistrar.java]: no, this is on purpose to be temporally backward compatible until all static tables are moved to the new approach.

File: sql/src/main/java/io/crate/metadata/table/StaticTableInfo.java [hunk 2]
```diff
@@ -59,6 +61,10 @@ public StaticTableInfo(RelationName ident, ColumnRegistrar columnRegistrar, List
         this(ident, columnRegistrar.infos(), columnRegistrar.columns(), primaryKey);
     }
 
+    public StaticTableInfo(RelationName ident, ColumnRegistrar columnRegistrar, String... primaryKey) {
```
Comment (kovrus) [inline sql/src/main/java/io/crate/metadata/table/StaticTableInfo.java]: should it be smth like `ColumnRegistrar<?>`

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 
 package io.crate.integrationtests;
 
+import com.carrotsearch.randomizedtesting.annotations.Seed;
```
Comment (kovrus) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: remove

File: sql/src/main/java/io/crate/execution/engine/collect/NestableCollectExpression.java [hunk 4]
```diff
@@ -23,39 +23,33 @@
...
 
 /**
  * Base interface for row based expressions.
  *
- * @param <TReturnValue> The returnType of the expression
+ * @param <R> The returnType of the expression
  */
-public abstract class NestableCollectExpression<TRow, TReturnValue>
-    implements CollectExpression<TRow, TReturnValue>, NestableInput<TReturnValue> {
+public interface NestableCollectExpression<T, R> extends CollectExpression<T, R>, NestableInput<R> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/NestableCollectExpression.java]: The changes here in this class look unrelated to the ColumnRegistry changes. Can we revert them / split them out into a separate PR or commit?

Comment (mkleen) [reply sql/src/main/java/io/crate/execution/engine/collect/NestableCollectExpression.java]: yes, sure.

File: sql/src/main/java/io/crate/metadata/table/StaticTableInfo.java (hunk 2 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/metadata/table/StaticTableInfo.java]: I don't think we win anything here if we add a generic type. I would not bind StaticTableInfo to T of ColumnRegistrar and it is also does not make sense to bind only the constructor. It would not increase type safety from my perspective.

File: common/src/main/java/io/crate/types/DataTypes.java [hunk 5]
```diff
@@ -77,6 +77,7 @@
     public static final DataType STRING_ARRAY = new ArrayType(STRING);
     public static final DataType INTEGER_ARRAY = new ArrayType(INTEGER);
     public static final DataType SHORT_ARRAY = new ArrayType(SHORT);
+    public static final DataType LONG_ARRAY = new ArrayType(LONG);
```
Comment (kovrus) [inline common/src/main/java/io/crate/types/DataTypes.java]: seems not to be used anywhere

Comment (mkleen) [reply common/src/main/java/io/crate/types/DataTypes.java]: well spotted, thanks.

File: sql/src/main/java/io/crate/metadata/sys/SysSnapshotsTableInfo.java [hunk 6]
```diff
@@ -55,36 +54,24 @@
...
-        super(IDENT, new ColumnRegistrar(IDENT, GRANULARITY)
-                .register(Columns.NAME, DataTypes.STRING)
-                .register(Columns.REPOSITORY, DataTypes.STRING)
-                .register(Columns.CONCRETE_INDICES, new ArrayType(DataTypes.STRING))
-                .register(Columns.STARTED, DataTypes.TIMESTAMPZ)
-                .register(Columns.FINISHED, DataTypes.TIMESTAMPZ)
-                .register(Columns.VERSION, DataTypes.STRING)
-                .register(Columns.STATE, DataTypes.STRING),
-            PRIMARY_KEY);
+        super(IDENT, columnRegistrar(),"name","repository");
```
Comment (kovrus) [inline sql/src/main/java/io/crate/metadata/sys/SysSnapshotsTableInfo.java]: spaces

File: sql/src/main/java/io/crate/metadata/sys/SysHealthTableInfo.java [hunk 7]
```diff
@@ -34,56 +32,37 @@
...
-            new ColumnRegistrar(IDENT, GRANULARITY)
-                .register(Columns.TABLE_NAME, DataTypes.STRING)
-                .register(Columns.TABLE_SCHEMA, DataTypes.STRING)
-                .register(Columns.PARTITION_IDENT, DataTypes.STRING)
-                .register(Columns.HEALTH, DataTypes.STRING)
-                .register(Columns.SEVERITY, DataTypes.SHORT)
-                .register(Columns.MISSING_SHARDS, DataTypes.LONG)
-                .register(Columns.UNDERREPLICATED_SHARDS, DataTypes.LONG),
-            Collections.emptyList());
+        super(IDENT,columnRegistrar());
```
Comment (kovrus) [inline sql/src/main/java/io/crate/metadata/sys/SysHealthTableInfo.java]: space

**Comment by mkleen** [comment]
retest this please
